Mathlib/Analysis/NormedSpace/MultipliableUniformlyOn.lean

```diff
@@ -0,0 +1,130 @@
+/-
+Copyright (c) 2025 Chris Birkbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birkbeck
+-/
+import Mathlib.Algebra.Order.Ring.Star
+import Mathlib.Analysis.CStarAlgebra.Classes
+import Mathlib.Analysis.NormedSpace.FunctionSeries
+import Mathlib.Analysis.SpecialFunctions.Log.Summable
+import Mathlib.Topology.Algebra.InfiniteSum.UniformOn
+
+/-!
+# Uniform convergence of products of functions
+
+We gather some results about the uniform convergence of infinite products, in particular those of
+the form `∏' i, (1 + f i x)` for a sequence `f` of complex valued functions.
+-/
+
+open Filter Function Complex Finset Topology
+
+variable {α β ι : Type*} [UniformSpace β] [AddGroup β] [IsUniformAddGroup β] [LinearOrder β]
+  [OrderTopology β] [AddLeftMono β] [AddRightMono β]
+section cexp_clog
+
+/- These next three results feel like they should go elsewhere, but find_home says to leave them
+here. -/
+lemma TendstoUniformlyOn.eventually_forall_lt {f : ι → α → β} {p : Filter ι} {g : α → β}
+    {K : Set α} {u v : β} (huv : u < v) (hf : TendstoUniformlyOn f g p K) (hg : ∀ x ∈ K, g x ≤ u) :
+    ∀ᶠ i in p, ∀ x ∈ K, f i x < v := by
+  simp only [tendstoUniformlyOn_iff_tendsto, uniformity_eq_comap_neg_add_nhds_zero,
+    tendsto_iff_eventually, eventually_comap, Prod.forall] at *
+  conv at hf => enter [2]; rw [eventually_iff_exists_mem]
+  have hf2 := hf (fun x ↦ -x.1 + x.2 < -u + v) ⟨_, (isOpen_gt' (-u + v)).mem_nhds (by simp [huv]),
+    fun y hy a b hab ↦ (hab.symm ▸ hy :)⟩
+  filter_upwards [eventually_prod_principal_iff.mp hf2] with i hi x hx
+  simpa using add_lt_add_of_le_of_lt (hg x hx) (hi x hx)
+
+lemma TendstoUniformlyOn.eventually_forall_le {f : ι → α → β} {p : Filter ι} {g : α → β}
+    {K : Set α} {u v : β} (huv : u < v) (hf : TendstoUniformlyOn f g p K) (hg : ∀ x ∈ K, g x ≤ u) :
+    ∀ᶠ i in p, ∀ x ∈ K, f i x ≤ v := by
+  filter_upwards [hf.eventually_forall_lt huv hg] with i hi x hx using (hi x hx).le
+
+lemma TendstoUniformlyOn.comp_cexp {p : Filter ι} {f : ι → α → ℂ} {g : α → ℂ}
+    {K : Set α} (hf : TendstoUniformlyOn f g p K) (hg : BddAbove <| (fun x ↦ (g x).re) '' K) :
+    TendstoUniformlyOn (cexp ∘ f ·) (cexp ∘ g) p K := by
+  obtain ⟨v, hv⟩ : ∃ v, ∀ x ∈ K, (g x).re ≤ v := hg.imp fun _ h ↦ by simpa [mem_upperBounds] using h
+  have : ∀ᶠ i in p, ∀ x ∈ K, (f i x).re ≤ v + 1 := hf.re.eventually_forall_le (lt_add_one v) hv
+  refine (UniformContinuousOn.cexp _).comp_tendstoUniformlyOn_eventually (by simpa) ?_ hf
+  simpa using fun a ha ↦ (hv a ha).trans (lt_add_one v).le
```

Comment:
```suggestion
  exact fun a ha ↦ (hv a ha).trans (lt_add_one v).le
```
(We should have a linter that checks automatically if `simpa using foo` can become `exact foo`.)